#' dcast to a matrix
#'
#' Reformat a long data.table to a wide matrix, preserving rownames and type
#'
#' @param table The \code{data.table} to dcast.
#' @param ... Extra arguments to be passed to \code{dcast.data.table}
#'
#' @export
dcast2mat <- function(table, ...){
  # dcast the table
  dat <- dcast.data.table(data = table, ...)
  rn <- dat[[1]]
  dat <- as.matrix(dat[, -1])
  rownames(dat) <- rn
  return(dat)
  # extract rownames and column names
  # cast as matrix | this must be done after removing the rownames column to prevent additiohnal coercion
  # set the correct type
}


# @examples
# idx <- sample(1:5, 12, replace = t)
# dt1 <- data.table(original = letters[11:15][idx], expected = letters[idx])
# dt_match(dt1, letters[11:15], letters[1:5])


#' Process methylation data from bismark
#'
#' Read output from bismark and return methylation states for all reads/positions.
#'
#' @param prefix A \code{character}. The prefix for all file.
#' @param path A \code{character}. The path with the bam files generated by
#'  \code{bismark}.
#'
#' @return  A \code{data.table} with one row per position per read and methylation
#' state.
#' @export
bismark_results <- function(prefix, path = "."){
  lf <- list.files(path = path, pattern = prefix)
  cpg_obf <- grep("cpg_ob", lf, value = t)
  cpg_otf <- grep("cpg_ot", lf, value = t)
  if(length(cpg_obf) != 1 | length(cpg_otf) != 1){
    lobf <- length(cpg_obf)
    lotf <- length(cpg_otf)
    stop(paste0("You should have exactly 1 ob file and 1 ot file. you have ", lobf, " and ", lotf, "."))
  }
  cpg_ob <- fread(cpg_obf, header = f)
  cpg_ot <- fread(cpg_otf, header = f)
  cpg_ob[, Mstrand := "ob"]
  cpg_ot[, Mstrand := "ot"]
  cpg <- rbind(cpg_ob, cpg_ot)
  setnames(cpg, c("id", "meth_state", "chr", "pos", "meth_call", "strand"))
  return(cpg)
}
